src/ray/core_worker/task_manager.cc

```diff
@@ -48,18 +49,39 @@ void TaskManager::CompletePendingTask(const TaskID &task_id,
   }
 }
 
-void TaskManager::FailPendingTask(const TaskID &task_id, rpc::ErrorType error_type) {
+void TaskManager::PendingTaskFailed(const TaskID &task_id, rpc::ErrorType error_type) {
   RAY_LOG(DEBUG) << "Failing task " << task_id;
-  int64_t num_returns;
+  int num_retries_left = 0;
+  TaskSpecification spec;
   {
     absl::MutexLock lock(&mu_);
     auto it = pending_tasks_.find(task_id);
     RAY_CHECK(it != pending_tasks_.end())
         << "Tried to complete task that was not pending " << task_id;
-    num_returns = it->second;
-    pending_tasks_.erase(it);
+    spec = it->second.first;
+    num_retries_left = it->second.second;
+    if (num_retries_left == 0) {
```

Comment:
<= 0 for safety

Reply:
I wanted to reserve -1 for infinite retries later on, but we can add a check here (the count should never go below 0).